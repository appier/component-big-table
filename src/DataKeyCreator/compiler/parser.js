const parser = (tokens) => {
  let position = 0;

  const additive = () => {
    let left = multiplicative();
    if(position < tokens.length){
      let token = tokens[position];
      // console.log('additive', position, token)
      while ( token && (token.value === '+' || token.value === '-' ) ) {
        position++;
        left = {
          type: 'BinaryExpression',
          left: left,
          operator: token.value,
          right: multiplicative()
        };
        token = tokens[position];
      }
    }
    return left;
  };

  const multiplicative = () => {
    let left = primay();
    if(position < tokens.length){
      let token = tokens[position];
      // console.log('multiplicative', position, token)
      while ( token && (token.value === '*' || token.value === '/') ) {
        position++;
        left = {
          type: 'BinaryExpression',
          operator: token.value,
          left: left,
          right: primay(),
        };
        token = tokens[position];
      }
    }
    return left;
  }

  const primay = () => {
    if(position < tokens.length){
      let token = tokens[position];
      let ret = null;
      // console.log('primay', position, token)
      if ( token.value === '(' ) {
        position++;
        ret = additive();
        token = tokens[position];
        // consume ')'
        if(token && token.value === ')') {
          position++;
        }
        else {
          throw 'Unmatched paren: it seems you lack a ) in expression';
        }
      }
      else if ( token.type === 'number' ) {
        position++;
        ret = {
          type: 'NumberLiteral',
          value: token.value,
        };
      }
      else if ( token.type === 'identifier' ) {
        position++;
        ret = {
          type: 'Identifier',
          value: token.value,
        };
      }
      else {
        throw 'Unexpected token: ' + token.value;
      }
      return ret;
    }
  }

  let body = [];
  while (position < tokens.length) {
    body.push(additive());
  }

  let ast = {
    type: 'Program',
    body: body,
  };

  return ast;
}

export default parser;