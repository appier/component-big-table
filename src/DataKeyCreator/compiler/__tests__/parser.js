import parser from '../parser';

describe('Parser', () => {

  it('identifier', () => {
    expect(parser(
      [
        { type: 'identifier', value: 'cost' }
      ]
    )).toMatchSnapshot();
  });

  it('BinaryExpression: identifier', () => {
    expect(parser(
      [
        { type: 'identifier', value: 'action' },
        { type: 'operator', value: '/' },
        { type: 'identifier', value: 'cost' }
      ]
    )).toMatchSnapshot();
  });

  it('BinaryExpression: number', () => {
    expect(parser(
      [
        { type: 'number', value: '1' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '2' }
      ]
    )).toMatchSnapshot();
  });

  it('BinaryExpression: should able to chain', () => {
    expect(parser(
      [
        { type: 'number', value: '1' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '3' },
      ]
    )).toMatchSnapshot();
  });

  it('BinaryExpression: operator should have precedence', () => {
    // 1 + 2 * 3
    expect(parser(
      [
        { type: 'number', value: '1' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '3' }
      ]
    )).toMatchSnapshot();
  });

  it('BinaryExpression: operator should have precedence 2', () => {
    // 1 * 2 + 3 * 4
    expect(parser(
      [
        { type: 'number', value: '1' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '3' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '4' },
      ]
    )).toMatchSnapshot();
  });


  it('BinaryExpression: operator should have precedence 3', () => {
     // 1 + 2 * 3 - 4 / 5 => 1 + ( 2 * 3 ) - ( 4 / 5 )
    expect(parser(
      [
        { type: 'number', value: '1' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '*' },
        { type: 'number', value: '3' },
        { type: 'operator', value: '-' },
        { type: 'number', value: '4' },
        { type: 'operator', value: '/' },
        { type: 'number', value: '5' }
      ]
    )).toMatchSnapshot();
  });

  it('Parn Handling: basic', () => {
     // ( 1 + 2 )
    expect(parser(
      [
        { type: 'paren', value: '(' },
        { type: 'number', value: '1' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '2' },
        { type: 'paren', value: ')' },
      ]
    )).toMatchSnapshot();
  });

  it('Parn Handling: paren precedence', () => {
     // 1 * ( 2 + 3 )
    expect(parser(
      [
        { type: 'number', value: '1' },
        { type: 'operator', value: '*' },
        { type: 'paren', value: '(' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '3' },
        { type: 'paren', value: ')' },
      ]
    )).toMatchSnapshot();
  });

  it('Parn Handling: paren precedence 2', () => {
     // ( 1 + 2 )  * ( 2 + 3 )
    expect(parser(
      [
        { type: 'paren', value: '(' },
        { type: 'number', value: '1' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '2' },
        { type: 'paren', value: ')' },
        { type: 'operator', value: '*' },
        { type: 'paren', value: '(' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '3' },
        { type: 'paren', value: ')' },
      ]
    )).toMatchSnapshot();
  });

  it('Parn Handling: nested paren', () => {
     // ( ( 1 + 2 ) + 3)
    expect(parser(
      [
        { type: 'paren', value: '(' },
        { type: 'paren', value: '(' },
        { type: 'number', value: '1' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '2' },
        { type: 'paren', value: ')' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '3' },
        { type: 'paren', value: ')' },
      ]
    )).toMatchSnapshot();
  });

  it('Parn Handling: nested paren 2', () => {
     // ( 1 + ( 2 + 3 ) )
    expect(parser(
      [
        { type: 'paren', value: '(' },
        { type: 'number', value: '1' },
        { type: 'operator', value: '+' },
        { type: 'paren', value: '(' },
        { type: 'number', value: '2' },
        { type: 'operator', value: '+' },
        { type: 'number', value: '3' },
        { type: 'paren', value: ')' },
        { type: 'paren', value: ')' },
      ]
    )).toMatchSnapshot();
  });

  it('Parn Error: no close paren', () => {
     // ( 1 + 2
    try {
      expect(parser(
        [
          { type: 'paren', value: '(' },
          { type: 'number', value: '1' },
          { type: 'operator', value: '+' },
          { type: 'number', value: '2' },
        ]
      ))
    } catch (err) {
      expect(String(err)).toMatchSnapshot();
    }
  });

  it('Parn Error: no close paren', () => {
     // ( 1 + 2 * ( 2 + 3 )
    try {
      expect(parser(
        [
          { type: 'paren', value: '(' },
          { type: 'number', value: '1' },
          { type: 'operator', value: '+' },
          { type: 'number', value: '2' },
          { type: 'operator', value: '*' },
          { type: 'paren', value: '(' },
          { type: 'number', value: '2' },
          { type: 'operator', value: '+' },
          { type: 'number', value: '3' },
          { type: 'paren', value: ')' },
        ]
      ))
    } catch (err) {
      expect(String(err)).toMatchSnapshot();
    }
  });


});

