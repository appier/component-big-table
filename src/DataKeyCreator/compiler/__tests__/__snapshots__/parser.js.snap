exports[`Parser BinaryExpression: identifier 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "type": "Identifier",
        "value": "action",
      },
      "operator": "/",
      "right": Object {
        "type": "Identifier",
        "value": "cost",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser BinaryExpression: number 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "type": "NumberLiteral",
        "value": "1",
      },
      "operator": "*",
      "right": Object {
        "type": "NumberLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser BinaryExpression: operator should have precedence 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "type": "NumberLiteral",
        "value": "1",
      },
      "operator": "+",
      "right": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "operator": "*",
        "right": Object {
          "type": "NumberLiteral",
          "value": "3",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser BinaryExpression: operator should have precedence 2 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        "operator": "*",
        "right": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "type": "BinaryExpression",
      },
      "operator": "+",
      "right": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "3",
        },
        "operator": "*",
        "right": Object {
          "type": "NumberLiteral",
          "value": "4",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser BinaryExpression: operator should have precedence 3 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        "operator": "+",
        "right": Object {
          "left": Object {
            "type": "NumberLiteral",
            "value": "2",
          },
          "operator": "*",
          "right": Object {
            "type": "NumberLiteral",
            "value": "3",
          },
          "type": "BinaryExpression",
        },
        "type": "BinaryExpression",
      },
      "operator": "-",
      "right": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "4",
        },
        "operator": "/",
        "right": Object {
          "type": "NumberLiteral",
          "value": "5",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser BinaryExpression: should able to chain 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        "operator": "*",
        "right": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "type": "BinaryExpression",
      },
      "operator": "*",
      "right": Object {
        "type": "NumberLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser Parn Error: no close paren 1`] = `"Unmatched paren: it seems you lack a ) in expression"`;

exports[`Parser Parn Error: no close paren 2`] = `"Unmatched paren: it seems you lack a ) in expression"`;

exports[`Parser Parn Handling: basic 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "type": "NumberLiteral",
        "value": "1",
      },
      "operator": "+",
      "right": Object {
        "type": "NumberLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser Parn Handling: nested paren 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        "operator": "+",
        "right": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "type": "BinaryExpression",
      },
      "operator": "+",
      "right": Object {
        "type": "NumberLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser Parn Handling: nested paren 2 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "type": "NumberLiteral",
        "value": "1",
      },
      "operator": "+",
      "right": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "operator": "+",
        "right": Object {
          "type": "NumberLiteral",
          "value": "3",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser Parn Handling: paren precedence 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "type": "NumberLiteral",
        "value": "1",
      },
      "operator": "*",
      "right": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "operator": "+",
        "right": Object {
          "type": "NumberLiteral",
          "value": "3",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser Parn Handling: paren precedence 2 1`] = `
Object {
  "body": Array [
    Object {
      "left": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "1",
        },
        "operator": "+",
        "right": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "type": "BinaryExpression",
      },
      "operator": "*",
      "right": Object {
        "left": Object {
          "type": "NumberLiteral",
          "value": "2",
        },
        "operator": "+",
        "right": Object {
          "type": "NumberLiteral",
          "value": "3",
        },
        "type": "BinaryExpression",
      },
      "type": "BinaryExpression",
    },
  ],
  "type": "Program",
}
`;

exports[`Parser identifier 1`] = `
Object {
  "body": Array [
    Object {
      "type": "Identifier",
      "value": "cost",
    },
  ],
  "type": "Program",
}
`;
